### PR TITLE
Drop Python 3.6, set up pre-commit, switch to Hatch, etc.

### PR DESCRIPTION
**Pull request by akx**
Yuletide cleanup 🎄 🧹

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      - id: isort
+  - repo: https://github.com/PyCQA/flake8
+    rev: 6.0.0
+    hooks:
+      - id: flake8
+        additional_dependencies: &flake8_dependencies
+          - flake8-annotations
+          - flake8-bugbear
+          - flake8-docstrings
+          - flake8-isort
```
Comment (ruksi) [inline .pre-commit-config.yaml:21]: is there a reason to have **both** `isort` (introduced through the new pre-hooks) and `flake8-isort` (was previously used as the `isort` wrapper)?

_could_ theoretically cause an issue that `isort` is different version of `flake8-isort` and thus they keep wrangling imports to their desired order to infinity (not a huge risk as `isort` is quite stable)

File: valohai_yaml/utils/merge.py [hunk 2]
```diff
@@ -19,7 +19,7 @@ def merge_dicts(
     out: TD = type(a)()
 
     # Hack to keep the iteration order the same...
-    keys = list(a)  # type: List[Any]
+    keys: List[Any] = list(a)
```
Comment (ruksi) [inline valohai_yaml/utils/merge.py:22]: probably belongs to the previous commit but no biggie

**Review by ruksi — CHANGES_REQUESTED** [review]
one question about `isort`

File: .pre-commit-config.yaml (hunk 1 above)
Comment (akx) [reply .pre-commit-config.yaml:21]: Oh, yeah, good copypaste catch.

**Review by ruksi — APPROVED** [review]
looking stellar 🌃

**Pull request merged**